File: embedded/ground_station_esp32s3/src/telemetry.cpp
#include "ground_station/telemetry.hpp"

#include <cmath>
#include <cstddef>
#include <cstdint>
#include <limits>

namespace ground_station {
namespace {

constexpr double kFreshnessToleranceS = 1.0e-12;

std::int32_t read_le_i32(const Bytes& bytes, const std::size_t offset) {
    const auto value = static_cast<std::uint32_t>(bytes[offset]) |
                       (static_cast<std::uint32_t>(bytes[offset + 1U]) << 8U) |
                       (static_cast<std::uint32_t>(bytes[offset + 2U]) << 16U) |
                       (static_cast<std::uint32_t>(bytes[offset + 3U]) << 24U);
    const auto signed_value = value <= 0x7FFFFFFFU
                                  ? static_cast<std::int64_t>(value)
                                  : static_cast<std::int64_t>(value) - 0x100000000LL;
    return static_cast<std::int32_t>(signed_value);
}

std::int16_t read_le_i16(const Bytes& bytes, const std::size_t offset) {
    const std::uint16_t value = static_cast<std::uint16_t>(
        static_cast<std::uint16_t>(bytes[offset]) |
        static_cast<std::uint16_t>(static_cast<std::uint16_t>(bytes[offset + 1U]) << 8U));
    const auto signed_value = value <= 0x7FFFU
                                  ? static_cast<std::int32_t>(value)
                                  : static_cast<std::int32_t>(value) - 0x10000;
    return static_cast<std::int16_t>(signed_value);
}

bool is_fresh(const double age_s, const double limit_s) {
    return age_s >= -kFreshnessToleranceS && age_s <= limit_s + kFreshnessToleranceS;
}

}  // namespace

TelemetryCache::TelemetryCache(const FreshnessPolicy policy) : policy_(policy) {}

void TelemetryCache::ingest(const V7Frame& frame, const double steady_now_s) {
    ++link_sequence_;
    last_link_s_ = steady_now_s;

    if (frame.id == 0x08U && frame.data.size() >= 8U) {
        ++position_sequence_;
        position_ = PositionSample{position_sequence_, steady_now_s, read_le_i32(frame.data, 0U),
                                   read_le_i32(frame.data, 4U), 0.0, true};
        return;
    }
    if (frame.id == 0x06U && frame.data.size() >= 2U) {
        ++status_sequence_;
        status_ = StatusSample{status_sequence_, steady_now_s, frame.data[0], frame.data[1] == 1U,
                               0.0, true};
        return;
    }
    if (frame.id == 0x51U && frame.data.size() >= 2U) {
        ++diagnostic_sequence_;
        Diagnostic51 diagnostic;
        diagnostic.sequence = diagnostic_sequence_;
        diagnostic.received_s = steady_now_s;
        diagnostic.mode = frame.data[0];
        diagnostic.state = frame.data[1];
        if (frame.data.size() >= 5U) {
            diagnostic.quality = frame.data.back();
        }
        if (diagnostic.mode == 2U && frame.data.size() >= 15U) {
            diagnostic.integrated_x_cm = read_le_i16(frame.data, 10U);
            diagnostic.integrated_y_cm = read_le_i16(frame.data, 12U);
            diagnostic.quality = frame.data[14];
        }
        diagnostic_51_ = diagnostic;
    }
}

TelemetrySnapshot TelemetryCache::snapshot(const double steady_now_s) const {
    TelemetrySnapshot result;
    result.position = position_;
    if (result.position.has_value()) {
        result.position->age_s = steady_now_s - result.position->received_s;
        result.position->fresh = is_fresh(result.position->age_s, policy_.position_max_age_s);
    }
    result.status = status_;
    if (result.status.has_value()) {
        result.status->age_s = steady_now_s - result.status->received_s;
        result.status->fresh = is_fresh(result.status->age_s, policy_.status_max_age_s);
    }
    result.diagnostic_51 = diagnostic_51_;
    result.link.sequence = link_sequence_;
    result.link.received_s = last_link_s_;
    if (last_link_s_.has_value()) {
        result.link.age_s = steady_now_s - *last_link_s_;
        result.link.fresh = is_fresh(result.link.age_s, policy_.link_max_age_s);
    } else {
        result.link.age_s = std::numeric_limits<double>::infinity();
    }
    return result;
}

}  // namespace ground_station

File: ros2_ws/src/ed_uav_fcu_bridge/ed_uav_fcu_bridge/telemetry.py
"""Freshness-aware telemetry cache for native V7 input frames."""

from __future__ import annotations

import struct
from dataclasses import dataclass, replace
from math import isclose

from .v7_codec import FrameDecodeError, V7Frame, decode_frame


@dataclass(frozen=True, slots=True)
class FreshnessPolicy:
    """Steady-clock freshness limits for the bridge's telemetry sources."""

    position_max_age_s: float = 0.20
    aux_status_max_age_s: float = 0.50
    link_max_age_s: float = 0.50


@dataclass(frozen=True, slots=True)
class PositionSample:
    source_sequence: int
    received_steady_s: float
    forward_m: float
    right_m: float
    valid: bool
    steady_age_s: float
    source_stamp_ns: int | None


@dataclass(frozen=True, slots=True)
class StatusSample:
    source_sequence: int
    received_steady_s: float
    mode: int
    motors_armed: bool
    valid: bool
    steady_age_s: float
    source_stamp_ns: int | None


@dataclass(frozen=True, slots=True)
class AuxSample:
    source_sequence: int
    received_steady_s: float
    channels_us: tuple[int, ...]
    valid: bool
    steady_age_s: float
    source_stamp_ns: int | None

    @property
    def aux1_us(self) -> int:
        """Return AUX1, the fifth channel in the complete 0x40 payload."""
        return self.channels_us[4]

    @property
    def aux6_us(self) -> int:
        """Return AUX6, preserving the existing mission-start switch mapping."""
        return self.channels_us[9]


@dataclass(frozen=True, slots=True)
class FlowDiagnostic51:
    source_sequence: int
    received_steady_s: float
    mode: int
    state: int
    quality: int | None
    integrated_x_cm: int | None
    integrated_y_cm: int | None
    source_stamp_ns: int | None


@dataclass(frozen=True, slots=True)
class LinkSample:
    source_sequence: int
    received_steady_s: float | None
    valid: bool
    steady_age_s: float


@dataclass(frozen=True, slots=True)
class TelemetrySnapshot:
    position: PositionSample | None
    status: StatusSample | None
    aux: AuxSample | None
    flow_diagnostic: FlowDiagnostic51 | None
    altitude_m: float | None
    battery_voltage_v: float | None
    link: LinkSample


class TelemetryCache:
    """Mutable V7 cache; only ID 0x08 is allowed to mutate continuous position."""

    def __init__(self, policy: FreshnessPolicy) -> None:
        self.policy = policy
        self._position: PositionSample | None = None
        self._status: StatusSample | None = None
        self._aux: AuxSample | None = None
        self._flow_diagnostic: FlowDiagnostic51 | None = None
        self._altitude_m: float | None = None
        self._battery_voltage_v: float | None = None
        self._position_sequence = 0
        self._status_sequence = 0
        self._aux_sequence = 0
        self._diagnostic_sequence = 0
        self._link_sequence = 0
        self._last_link_steady_s: float | None = None

    def ingest_raw(self, raw: bytes, steady_now: float, source_stamp_ns: int | None = None) -> bool:
        """Decode and cache one raw V7 frame, returning False for malformed input."""
        try:
            frame = decode_frame(raw)
        except FrameDecodeError:
            return False
        self.ingest_frame(frame, steady_now, source_stamp_ns)
        return True

    def ingest_frame(self, frame: V7Frame, steady_now: float, source_stamp_ns: int | None = None) -> None:
        """Cache one checksum-verified V7 frame with receive-time provenance."""
        self._link_sequence += 1
        self._last_link_steady_s = steady_now
        match frame.frame_id:
            case 0x08 if len(frame.data) >= 8:
                self._position_sequence += 1
                forward_cm, right_cm = struct.unpack_from("<ii", frame.data)
                self._position = PositionSample(
                    self._position_sequence,
                    steady_now,
                    forward_cm / 100.0,
                    right_cm / 100.0,
                    True,
                    0.0,
                    source_stamp_ns,
                )
            case 0x51 if len(frame.data) >= 2:
                self._cache_flow_diagnostic(frame.data, steady_now, source_stamp_ns)
            case 0x06 if len(frame.data) >= 2:
                self._status_sequence += 1
                self._status = StatusSample(
                    self._status_sequence,
                    steady_now,
                    frame.data[0],
                    frame.data[1] == 1,
                    True,
                    0.0,
                    source_stamp_ns,
                )
            case 0x40 if len(frame.data) >= 20:
                self._aux_sequence += 1
                channels_us = struct.unpack_from("<10h", frame.data)
                self._aux = AuxSample(
                    self._aux_sequence,
                    steady_now,
                    channels_us,
                    True,
                    0.0,
                    source_stamp_ns,
                )
            case 0x05 if len(frame.data) >= 4:
                self._altitude_m = struct.unpack_from("<i", frame.data)[0] / 100.0
            case 0x0D if len(frame.data) >= 2:
                self._battery_voltage_v = struct.unpack_from("<H", frame.data)[0] / 100.0
            case _:
                pass

    def _cache_flow_diagnostic(self, data: bytes, steady_now: float, source_stamp_ns: int | None) -> None:
        self._diagnostic_sequence += 1
        mode, state = data[:2]
        match mode:
            case 2 if len(data) >= 15:
                integrated_x_cm, integrated_y_cm = struct.unpack_from("<hh", data, 10)
                quality: int | None = data[14]
            case _:
                integrated_x_cm = None
                integrated_y_cm = None
                quality = data[-1] if len(data) >= 5 else None
        self._flow_diagnostic = FlowDiagnostic51(
            self._diagnostic_sequence,
            steady_now,
            mode,
            state,
            quality,
            integrated_x_cm,
            integrated_y_cm,
            source_stamp_ns,
        )

    def has_fresh_start_switch(self, steady_now: float) -> bool:
        """Return true only for a fresh AUX6 mission-start switch assertion."""
        aux = self._with_aux_age(steady_now)
        return aux is not None and aux.valid and aux.aux6_us > 1700

    def snapshot(self, steady_now: float) -> TelemetrySnapshot:
        """Return source-separated state with validity derived from steady-clock age."""
        return TelemetrySnapshot(
            position=self._with_position_age(steady_now),
            status=self._with_status_age(steady_now),
            aux=self._with_aux_age(steady_now),
            flow_diagnostic=self._flow_diagnostic,
            altitude_m=self._altitude_m,
            battery_voltage_v=self._battery_voltage_v,
            link=self._link_snapshot(steady_now),
        )

    def _with_position_age(self, steady_now: float) -> PositionSample | None:
        if self._position is None:
            return None
        age = steady_now - self._position.received_steady_s
        return replace(self._position, valid=self._within(age, self.policy.position_max_age_s), steady_age_s=age)

    def _with_status_age(self, steady_now: float) -> StatusSample | None:
        if self._status is None:
            return None
        age = steady_now - self._status.received_steady_s
        return replace(self._status, valid=self._within(age, self.policy.aux_status_max_age_s), steady_age_s=age)

    def _with_aux_age(self, steady_now: float) -> AuxSample | None:
        if self._aux is None:
            return None
        age = steady_now - self._aux.received_steady_s
        return replace(self._aux, valid=self._within(age, self.policy.aux_status_max_age_s), steady_age_s=age)

    def _link_snapshot(self, steady_now: float) -> LinkSample:
        if self._last_link_steady_s is None:
            return LinkSample(self._link_sequence, None, False, float("inf"))
        age = steady_now - self._last_link_steady_s
        return LinkSample(self._link_sequence, self._last_link_steady_s, self._within(age, self.policy.link_max_age_s), age)

    @staticmethod
    def _within(age: float, limit: float) -> bool:
        return age < limit or isclose(age, limit, rel_tol=0.0, abs_tol=1e-12)
